src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -99,6 +104,166 @@ using JoystickBlacklist = std::vector<JoystickBlacklistEntry>;
 JoystickBlacklist joystickBlacklist;
 
 const DWORD directInputEventBufferSize = 32;
+
+// For XInput
+std::vector<bool> xInputSlots;
+bool              directInputNeedsInvalidation = false;
+
+struct XInputCleanupData
+{
+    VARIANT                           var{};
+    IWbemLocator*                     pIWbemLocator{};
+    IEnumWbemClassObject*             pEnumDevices{};
+    std::array<IWbemClassObject*, 20> pDevices{};
+    IWbemServices*                    pIWbemServices{};
+    BSTR                              bstrNamespace{};
+    BSTR                              bstrDeviceID{};
+    BSTR                              bstrClassName{};
+    HRESULT                           comInitResult{};
+
+    XInputCleanupData()
+    {
+        VariantInit(&var);
+    }
+
+    ~XInputCleanupData()
+    {
+        VariantClear(&var);
+
+        SysFreeString(bstrNamespace);
+        SysFreeString(bstrDeviceID);
+        SysFreeString(bstrClassName);
+
+        for (auto* device : pDevices)
+            if (device)
+                device->Release();
+
+        if (pEnumDevices)
+            pEnumDevices->Release();
+        if (pIWbemLocator)
+            pIWbemLocator->Release();
+        if (pIWbemServices)
+            pIWbemServices->Release();
+        if (SUCCEEDED(comInitResult))
+            CoUninitialize();
+    }
+};
+
+// NOLINTBEGIN(readability-identifier-naming)
+// keeping GUIDs and UUIDs consistent
+// Define CLSID_WbemLocator
+const CLSID CLSID_WbemLocator = {0x4590f811, 0x1d3a, 0x11d0, {0x89, 0x1f, 0x00, 0xaa, 0x00, 0x4b, 0x2e, 0x24}};
+
+// Define IID_IWbemLocator
+const IID IID_IWbemLocator = {0xdc12a687, 0x737f, 0x11cf, {0x88, 0x4d, 0x00, 0xaa, 0x00, 0x4b, 0x2e, 0x24}};
+// NOLINTEND(readability-identifier-naming)
+
+// Function pointer type for XInputGetState
+using XInputGetStateFunc = DWORD(WINAPI*)(DWORD dwUserIndex, XINPUT_STATE* pState);
+
+XInputGetStateFunc mXInputGetState = nullptr;
+
+// See also https://learn.microsoft.com/en-us/windows/win32/xinput/xinput-and-directinput?redirectedfrom=MSDN
+[[nodiscard]] BOOL isXInputDevice(const GUID* pGuidProductFromDirectInput)
```

Comment:
The DirectInput SDK sample caches this data instead of querying it for every device, that might also make sense?
https://github.com/walbourn/directx-sdk-samples/blob/main/DirectInput/Joystick/joystick.cpp#L288

Reply:
And this stackoverflow post has another way of getting the vid/pid from the XInput index:
https://stackoverflow.com/a/68879988/4928207

Reply:
Yes, but I dislike using undocumented APIs unless there's an exceptionally compelling usecase for it. 

In this case, it isn't compelling enough.

Reply:
You know what? Never mind! I think you're right, and I'm going to have to use this to get the right description for each device we find on XInput, because otherwise we have no idea which device is which.